app/auth.go

```diff
@@ -11,23 +11,63 @@
 //    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 //    See the License for the specific language governing permissions and
 //    limitations under the License.
+
 package app
 
 import (
 	"os"
 	"strings"
 
 	"github.com/mendersoftware/mender/client"
+	"github.com/mendersoftware/mender/conf"
 	"github.com/mendersoftware/mender/datastore"
 	dev "github.com/mendersoftware/mender/device"
 	"github.com/mendersoftware/mender/store"
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 )
 
+// Constants for auth manager request methods
+const (
+	MethodGetAuthToken        = "GET_AUTH_TOKEN"
+	MethodFetchAuthToken      = "FETCH_AUTH_TOKEN"
+	MethodForceFetchAuthToken = "FORCE_FETCH_AUTH_TOKEN"
+)
+
+const (
+	noAuthToken              = client.EmptyAuthToken
+	AuthManagerChannelBuffer = 1024
+)
+
+// AuthManagerRequest stores a request to the Mender authorization manager
+type AuthManagerRequest struct {
+	Sender string
+	Method string
+}
+
+// AuthManagerResponse stores a response from the Mender authorization manager
+type AuthManagerAck struct {
+	Method string
+	Error  error
+}
+
+// AuthManagerResponse stores a response from the Mender authorization manager
+type AuthManagerResponse struct {
+	Method    string
+	AuthToken client.AuthToken
+	Error     error
+}
+
+// AuthManager is the interface of a Mender authorization manager
 type AuthManager interface {
-	// returns true if authorization data is current and valid
-	IsAuthorized() bool
+	Bootstrap() menderError
+	ForceBootstrap()
+	GetInMessageChan() chan<- *AuthManagerRequest
+	GetAckMessageChan(name string) <-chan *AuthManagerAck
+	GetOutMessageChan(name string) <-chan *AuthManagerResponse
```

Comment:
Take this comment with a grain of salt, because I'm not sure if I fully understand everything with the channels.

I have a suggested modification of this approach which perhaps is more like a socket based client/server would operate: Instead of returning channels by passing a name, make the caller of `AuthManager` the owner of the return channels, and pass this channel in the `AuthManagerRequest` channel. Then the `AuthManager` returns the response on this channel only. This also addresses one of the other comments I made, namely that the response should not be returned on all channels, only to the requester. I think if this approach is used, you don't need an Ack channel, because all calls will be synchronous if you first post to the in-channel (request channel), and then wait on the response channel that you gave in `AuthManagerRequest`.

Then there is also the "broadcast" when a new token is available, but I think this should be completely separate from the request/response framework. Here it would be more appropriate to register listening channels. When a new token is available, this would be published on all registered channels.

Not sure if this was clear from the task description, but `FetchJwtToken` is also synchronous in the sense that it immediately returns "Ok, a fetch is scheduled", and it is the broadcast channel which is used to publish that a new token is available (so it needs to register a listening channel first if it plans to wait for it). So all calls are synchronous from a calling perspective (but their effects might not be).

Reply:
I updated the PR to follow the pattern you are suggesting: the caller passes in the chan for the responses, which is consumed immediately to get the response synchronously. I still have issues in understanding the difference between getAuthToken and FetchAuthToken. Let's have a chat and discuss the details.

Reply:
Reimplemented as discussed, following the pattern you outlined.